RDB Loader 5.2.0 (#191)
snowplow/documentation: docs/pipeline-components-and-applications/loaders-storage-targets/snowplow-rdb-loader/loading-transformed-data/rdb-loader-configuration-reference/index.md

| `roleArn`                  | Required if 'NoCreds' is chosen as load auth method.. AWS Role ARN allowing Redshift to load data from S3.                                                                                                                  |
| `schema`                   | Required. Redshift schema name, eg “atomic”.                                                                                                                                    |
| `username`                 | Required. DB user with permissions to load data.                                                                                                                                |
| `password`                 | Required. Password of DB user.                                                                                                                                                  |
| `maxError`                 | Optional. Configures the [Redshift MAXERROR load option](https://docs.aws.amazon.com/redshift/latest/dg/copy-parameters-data-load.html#copy-maxerror). The default is 10.       |

> | `loadAuthMethod.*`         | Optional, default method is `NoCreds`. Specifies the auth method to use with 'COPY' statement. |

Reviewer: ```suggestion
| `loadAuthMethod.*` (since 5.2.0) | Optional, default method is `NoCreds`. Specifies the auth method to use with the `COPY` statement. |
```

> | `loadAuthMethod.type`      | Required if `loadAuthMethod` section is included. Specifies the type of the auth method. The possible values are `NoCreds` and `TempCreds`. <br/><br/>With 'NoCreds', no credentials will be passed to 'COPY' statement. Instead, Redshift cluster needs to be configured AWS Role ARN allowing Redshift to load data from S3 and this Role ARN needs to be passed field 'roleArn' above. More information can be found [here](https://docs.aws.amazon.com/redshift/latest/dg/copy-usage_notes-access-permissions.html). <br/><br/>With 'TempCreds', temporary credentials will be created for every load operation and these temporary credentials will be passed to 'COPY' statement. |

Reviewer: ```suggestion
| `loadAuthMethod.type`      | Required if `loadAuthMethod` section is included. Specifies the type of the authentication method. The possible values are `NoCreds` and `TempCreds`. <br/><br/>With `NoCreds`, no credentials will be passed to the `COPY` statement. Instead, Redshift cluster needs to be configured with an AWS Role ARN that allows it to load data from S3. This Role ARN needs to be passed in the `roleArn` setting above. You can find more information [here](https://docs.aws.amazon.com/redshift/latest/dg/copy-usage_notes-access-permissions.html). <br/><br/>With `TempCreds`, temporary credentials will be created for every load operation and these temporary credentials will be passed to the `COPY` statement. |
```
Can we add any info on which method is recommended and why?

Author: Since we operating customer's pipelines and we don't have access to their Redshift instances directly, `TempCreds` method has some advantage for us since it has one less step to configure on customer side. However, for ordinary users, I don't think any method significant advantage over the other one. Therefore, I think there is no need to recommend any method.

> | `loadAuthMethod.roleArn`   | Required if `loadAuthMethod.type` is `TempCreds`.IAM role that is used while creating temporary credentials. Created credentials will allow to access resources specified in the given role. “s3:GetObject*”, “s3:ListBucket”, and “s3:GetBucketLocation” permissions for transformer output S3 bucket should be specified in the role.   

Reviewer: I find it very confusing that we have two different `roleArn` settings...

Author: I agree but there is not much thing to do. Default method is `NoCreds`. Maybe it is best to not document `TempCreds` at all. As I said in the previous comment, for ordinary users, no method has any significant advantage to one to another. New `TempCreds` method is slightly easier to configure but there is not much difference.

> | `loadAuthMethod.roleArn`   | Required if `loadAuthMethod.type` is `TempCreds`.IAM role that is used while creating temporary credentials. Created credentials will allow to access resources specified in the given role. “s3:GetObject*”, “s3:ListBucket”, and “s3:GetBucketLocation” permissions for transformer output S3 bucket should be specified in the role.   

Reviewer: ```suggestion
| `loadAuthMethod.roleArn`   | Required if `loadAuthMethod.type` is `TempCreds`. IAM role that is used while creating temporary credentials. Created credentials will allow Redshift to access resources specified in the given role. The role should include the following permissions: `s3:GetObject*`, `s3:ListBucket`, and `s3:GetBucketLocation`.  
```

Reviewer: Should we clarify that “resources specified in the given role” should be S3 buckets? (Or is it just 1 bucket?) Will any other resources need to be specified in this role under any circumstances?

Author: Nope, there will be no such case.

Author: > Should we clarify that “resources specified in the given role” should be S3 buckets?

Following part explains this actually. Wouldn't be better to leave it as it is ?

```
permissions for transformer output S3 bucket should be specified in the role.   
```

> | `loadAuthMethod.roleArn`   | Required if `loadAuthMethod.type` is `TempCreds`.IAM role that is used while creating temporary credentials. Created credentials will allow to access resources specified in the given role. `s3:GetObject*`, `s3:ListBucket`, and `s3:GetBucketLocation` permissions for transformer output S3 bucket should be specified in the role.

Reviewer: ```suggestion
| `loadAuthMethod.roleArn`   | Required if `loadAuthMethod.type` is `TempCreds`. IAM role that is used while creating temporary credentials. This role should allow access to the S3 bucket the transformer will write data to, with the following permissions: `s3:GetObject:*`, `s3:ListBucket`, and `s3:GetBucketLocation`.
```

Trying to simplify a bit and reduce passive voice which is hard to read.

| `transformedStage.*`                       | Required if 'NoCreds' is chosen as load auth method. Snowflake stage for transformed events.                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                         |

> | `transformedStage.name`                    | Required if `transformedStage` is included. The name of the stage.                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                   |

Reviewer: Where did this go?

Author: Since we've introduced [this feature](https://github.com/snowplow/snowplow-rdb-loader/issues/1059), stage location doesn't need to be passed manually to `transformedStage.location`. Therefore, it is enough to only pass stage name. It is possible to pass stage name as `transformedStage` directly like this `transformedStage: "stage_name"`. Therefore, I've removed `"transformedStage": { "name": "", "location": ""}` type config options. But, this type of configuration will still work. I just didn't want to continue to document this type of configuration.

Reviewer: But it still applies to older versions, so I think we need to keep it and add a commend like `(not needed since 5.2.0)`

Author: Wouldn't it bloat the configs too much ?

Reviewer: There must be some way for those still running 5.1.x, 5.0.x, etc, to configure their app... When we have a lot of deprecated config options, we can archive this version of configuration and create a new tree of pages for RDB Loader 10.0.0 or whatever version we’ll have by that time :)

> | `transformedStage`                       | Required if 'NoCreds' is chosen as load auth method. Snowflake stage for transformed events.                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                         |

Reviewer: ```suggestion
| `transformedStage`                       | Required if `NoCreds` is chosen as load auth method. Snowflake stage for transformed events.                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                         |
```

Reviewer: Here and below

| `loadAuthMethod.*`                         | Optional, default method is `NoCreds`. Specifies the auth method to use with `COPY INTO` statement. Note that `TempCreds` auth method doesn't work when data is loaded from GCS. |
| `loadAuthMethod.type`                      | Required if `loadAuthMethod` section is included. Specifies the type of the auth method. The possible values are `NoCreds` and `TempCreds`. <br/><br/>With `NoCreds`, no credentials will be passed to `COPY INTO` statement. Instead, `transformedStage` and `folderMonitoringStage` specified above will be used. More information can be found [here](https://docs.snowflake.com/en/user-guide/data-load-s3-config-storage-integration.html). <br/><br/>With `TempCreds`, temporary credentials will be created for every load operation and these temporary credentials will be passed to `COPY INTO` statement. |

> | `loadAuthMethod.roleArn`                   | Required if `loadAuthMethod.type` is `TempCreds`.IAM role that is used while creating temporary credentials. Created credentials will allow to access resources specified in the given role. List of necessary permissions needs to be given to role specified in [here](https://docs.snowflake.com/en/user-guide/data-load-s3-config-aws-iam-user.html).                                                                                                                                                                                                                                                            |

Reviewer: ```suggestion
| `loadAuthMethod.roleArn`                   | Required if `loadAuthMethod.type` is `TempCreds`. IAM role that is used while creating temporary credentials. Created credentials will allow to access resources specified in the given role. List of necessary permissions needs to be given to role specified in [here](https://docs.snowflake.com/en/user-guide/data-load-s3-config-aws-iam-user.html).                                                                                                                                                                                                                                                            |
```

| `loadAuthMethod.type`                      | Required if `loadAuthMethod` section is included. Specifies the type of the auth method. The possible values are `NoCreds` and `TempCreds`. <br/><br/>With `NoCreds`, no credentials will be passed to `COPY INTO` statement. Databricks cluster needs to have permission to access transformer output S3 bucket. More information can be found [here](https://docs.databricks.com/administration-guide/cloud-configurations/aws/instance-profiles.html). <br/><br/>With `TempCreds`, temporary credentials will be created for every load operation and these temporary credentials will be passed to `COPY INTO` statement. With this way, Databricks cluster doesn't need permission to access to transformer output S3 bucket. This access will be provided by temporary credentials. |

> | `loadAuthMethod.roleArn`                   | Required if `loadAuthMethod.type` is `TempCreds`. IAM role that is used while creating temporary credentials. Created credentials will allow to access resources specified in the given role. In our case, `s3:GetObject\*`, `s3:ListBucket`, and `s3:GetBucketLocation` permissions for transformer output S3 bucket should be specified in the role.                                                                                                                                                                                                                                                                                                                                                                                                                                    |

Reviewer: ```suggestion
| `loadAuthMethod.roleArn`                   | Required if `loadAuthMethod.type` is `TempCreds`. IAM role that is used while creating temporary credentials. This role should allow access to the S3 bucket the loader will read data from, with the following permissions: `s3:GetObject:*`, `s3:ListBucket`, and `s3:GetBucketLocation`.                                                                                                                                                                                                                                                                                                                                                                                                                                    |
```

| `password`                                 | Required. [Databricks access token](https://docs.databricks.com/dev-tools/api/latest/authentication.html). Can be plain text, read from the EC2 parameter store or GCP secret manager (see below).                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                           |

> | `eventsOptimizePeriod` | Optional. The default value is `2 days`. Optimize period per table, that will be used as predicate for the OPTIMIZE command. |

Reviewer: ```suggestion
| `eventsOptimizePeriod` | Optional. The default value is `2 days`. Optimize period per table, that will be used as predicate for the `OPTIMIZE` command. |
```

|--------------------------------------|-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| `schedules.*`                        | Optional. Tasks scheduled to run periodically.                                                                                                                                                                                                                                                                                                                         |

> | `schedules.noOperation.[*]`          | Optional. Array of objects which specifies no-operation windows. Periodic schedules to stop loading.                                                                                                                                                                                                                                                                                                   |

Reviewer: ```suggestion
| `schedules.noOperation.[*]`          | Optional. Array of objects which specifies no-operation windows during which periodically scheduled tasks (configured in this section) will not run.                                                                                                                                                                                                                                                    |
```

| `schedules.noOperation.[*].when`     | Cron expression with second granularity.                                                                                                                                                                                                                                                                                                                                                                  |
| `schedules.noOperation.[*].duration` | For how long the loader should be paused.                                                                                                                                                                                                                                                                                                                                                                 |

> | `schedules.optimizeEvents` | Optional. The default value is `"0 0 0 ? * *"`. Cron expression with second granularity that specifies the schedule to run periodic OPTIMIZE statement on event table. (Only for Databricks Loader) |

Reviewer: ```suggestion
| `schedules.optimizeEvents` | Optional. The default value is `"0 0 0 ? * *"`. Cron expression with second granularity that specifies the schedule to periodically run an `OPTIMIZE` statement on event table. (Only for Databricks Loader) |
```

> | `schedules.optimizeEvents` | Optional. The default value is `"0 0 0 ? * *"`. Cron expression with second granularity that specifies the schedule to run periodically an `OPTIMIZE` statement on event table. (Only for Databricks Loader) |

Reviewer: ```suggestion
| `schedules.optimizeEvents` | Optional. The default value is `"0 0 0 ? * *"` (i.e. every day at 00:00, JVM timezone). Cron expression with second granularity that specifies the schedule to run periodically an `OPTIMIZE` statement on event table. (Only for Databricks Loader) |
```

> | `schedules.optimizeManifest` | Optional. The default value is `"0 0 5 ? * *"`. Cron expression with second granularity that specifies the schedule to run periodic OPTIMIZE statement on manifest table. (Only for Databricks Loader) |

Reviewer: ```suggestion
| `schedules.optimizeManifest` | Optional. The default value is `"0 0 5 ? * *"`. Cron expression with second granularity that specifies the schedule to periodically run an `OPTIMIZE` statement on manifest table. (Only for Databricks Loader) |
```

> | `schedules.optimizeManifest` | Optional. The default value is `"0 0 5 ? * *"`. Cron expression with second granularity that specifies the schedule to run periodically an `OPTIMIZE` statement on manifest table. (Only for Databricks Loader) |

Reviewer: ```suggestion
| `schedules.optimizeManifest` | Optional. The default value is `"0 0 5 ? * *"` (i.e. every day at 05:00 AM, JVM timezone). Cron expression with second granularity that specifies the schedule to run periodically an `OPTIMIZE` statement on manifest table. (Only for Databricks Loader) |
```
